Converter: use streaming JSON parser to construct frames from loki/prometheus responses

ryantxu opened the pull request:
The backend prometheus conversion is quite heavyweight -- this introduces a streaming parser that will construct frames directly from the json reader.  It works for both loki and prometheus.

This PR simply adds an internal utility that can then be incorporated into other PRs

ryantxu committed:
add parser to grafana

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
add gosec comments

toddtreece commented on pkg/tsdb/prometheus/parser/parser.go:
```diff
@@ -0,0 +1,325 @@
+				for iter.ReadArray() {
+					t, v, err := readTimeValuePair(iter)
+					if err == nil {
+						timeField.Append(t)
```
i think it's more efficient to build arrays and pass them in to `data.NewField` since that avoids the `interface{}` conversion with each call to `Append`

ryantxu replied:
interesting -- is that true even with the 2nd copy of the array?  (I have avoided the NewField approach because it did the double allocation)  :man_shrugging:

toddtreece replied:
it seems to depend on the field type https://github.com/grafana/grafana-plugin-sdk-go/pull/456

gabor commented:
idea: maybe it's enough to only support success-responses.  if the http-status-code is 200 (well, `20x`), then the user can use this streaming parser library, otherwise should handle error-responses by hand. arguments:
- keeps the code here simpler
- they are simple structures, not a big deal to parse them
- some older Loki versions return error-responses as a `content-type:text/plain` string
- some http-errors can always come like http401 from an authentication-proxy or whatever, so maybe it's good to force the user to handle the non-200 responses manually

this is just an idea, no strong feelings here 😄

ryantxu commented:
re: error handling -- I don't feel too strongly.  But if we support the JSON > DataQueryResponse here... then it is one place.  If the error structures/warnings are simple it seems better to just do them here.

If the error is text/plain :thinking: -- what do we currently do?

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

gabor commented:
> re: error handling -- I don't feel too strongly. But if we support the JSON > DataQueryResponse here... then it is one place. If the error structures/warnings are simple it seems better to just do them here.
> 
> If the error is text/plain 🤔 -- what do we currently do?

hmm..so:
when JSON-error, it looks like this (`content-type=application/json; charset=utf-8`)
```
{
    "code": 400,
    "message": "parse error at line 1, col 1: syntax error: unexpected IDENTIFIER",
    "status": "error"
}
```

when text-plain error, it looks like this (`content-type=`text/plain; charset=utf-8`):
```
parse error at line 1, col 1: syntax error: unexpected IDENTIFIER
```

(fun fact: latest released Loki version has a bug, it's error messages are plain-text-format, but content-type says JSON)

currently on main-branch Loki-datasource is a frontend datasource, so error-handling is happening in typescript.
and in typescript there is generic code that normalizes errors at https://github.com/grafana/grafana/blob/main/public/app/core/services/backend_srv.ts#L317-L323

so later, in the loki-data-source, you can always just take `err.data.message`, because it will be always good.


i have a draft-PR where i implement the error-handling in Loki, it's here:
https://github.com/grafana/grafana/pull/44643, the error-handling in it is this: https://github.com/grafana/grafana/blob/gabor/loki-custom-http/pkg/tsdb/loki/api.go#L80-L104

basically:
- if the http-response-status-code is not `20x`
    - then i try to json-parse the body-bytes (i ignore the content-type), and if that succeeds and the `.message` attribute is a non-empty string, i return that text as the error-text
    - otherwise i return the whole message as the error-text

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
move out of prometheus folder

ryantxu committed:
renam form simple

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
add examples from gabor

ryantxu committed:
now with time parsing

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
force UTC

gabor commented:
i created a PR that updates the loki-datasource to use this streaming JSON parser at https://github.com/grafana/grafana/pull/45653 . with that change, the framing-tests in loki-datasource all pass 👍

gabor commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,337 @@
+			frame = data.NewFrame("", timeField, valueField)
+			for iter.ReadArray() {
+				for l2Field := iter.ReadObject(); l2Field != ""; l2Field = iter.ReadObject() {
+					switch l2Field {
```
any reason why this `switch` has no `default` (maybe with return-error)?

ryantxu replied:
added a return with notices

gabor commented:
i did one more test, this time i have code that handles all 3 loki-dat-types (matrix, vector, stream) using the loki-go-data-structures, and tried to switch to this streaming-parser.

the only incompatiblity is that currently streams generate an additional field in the dataframe (the nanosecond-timestamp-in-string). if i drop that field, the results pass my snapshots tests 👍

torkelo commented:
Nice!

gabor commented:
replying to myself :)

> the only incompatiblity is that currently streams generate an additional field in the dataframe (the nanosecond-timestamp-in-string). if i drop that field, the results pass my snapshots tests 👍

and now i see we need that field too :)

gabor commented:
after more thinking & writing code:
> replying to myself :)
> 
> > the only incompatiblity is that currently streams generate an additional field in the dataframe (the nanosecond-timestamp-in-string). if i drop that field, the results pass my snapshots tests 👍
> 
> and now i see we need that field too :)

even so, the nanosecond-precision can be extracted from the normal `time.Time` field too, so i think the string-with-time is not needed. (but maybe i'm missing something)

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
add notice for unknown key

nathanrodman committed:
Alerting: fix alert groups grouping (#45012)

* fix multiple non-grouped groupings

* drop duplicate alerts

* add test for multiple groups without labels

Co-authored-by: gillesdemey <gilles.de.mey@gmail.com>

ying-jeanne committed:
Middleware: Fix IPv6 host parsing in CSRF check (#45911)

- Also create tests for this middleware

Co-authored-by: Kyle Brandt <kyle@grafana.com>

CatPerry committed:
Jaeger: Show loader when search options havent yet loaded (#45936)

* Swap out Select component for AsyncSelect to Jaeger search panel

leeoniya committed:
StateTimeline: fix duration in tooltip (#45955)

- Fixes duration in StateTimeline appearing incorrectly when "merge consecutive values" is enabled.

leeoniya committed:
BarChart: fix single group rendering (#45953)

leeoniya committed:
Histogram: auto-skip x tick labels to avoid overlap (#45996)

Jguer committed:
Service Accounts: Link final components in service accounts detail page (#45929)

* ServiceAccounts: Delete/Disable service account from details page

* ServiceAccounts: capitalize viewable messages from UI

* ServiceAccounts: Link new update endpoint to details page

* ServiceAccounts: reimplement service account retrieve to include is_disabled and only target service accounts

* Cleanup styles

* Fix modal show

* ServiceAccounts: simplify handler functions

* Apply suggestions from code review

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

Co-authored-by: Clarity-89 <homes89@ukr.net>
Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

zuchka committed:
Transformations: Use asterisk for First non-null label (#45940)

kalleep committed:
Access control: Filter users and teams by read permissions (#45968)

* pass signed in user and filter based on permissions

tolzhabayev committed:
Switching to github.event.number as issue.number not always work (#46018)

tolzhabayev committed:
Making yarn.lock bump work (#46016)

renovate[bot] committed:
Update sentry-javascript monorepo to v6.18.1 (#45842)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

srclosson committed:
Chore: Add function documentation for new functions (Prometheus visual query editor) (#45745)

* add missing function documentation

* added trigonometric functions

torkelo committed:
Prometheus: Query builder ux tweaks (#46017)

torkelo committed:
Prometheus: Query builder operation docs popover improvements (#46006)

JohnnyQQQQ committed:
Revert "Alerting: Calculate diff for two AlertRules (#45877)" (#46023)

This reverts commit 4e19d7df6352b0dcbb680aeee00cebc97a90d937.

grafanabot committed:
ReleaseNotes: Updated changelog and release notes for 8.3.7 (#46028)

timlevett committed:
(GitHubActions) Add grafana-partners team label to team board (#45990)

achatterjee-grafana committed:
Docs update default.ini file description (#46036)

* remove confusing wording

* fixed broken alerting links

yuri-tceretian committed:
Alerting: Revert Revert "Alerting: Calculate diff for two AlertRules" (#46034)

* Revert "Revert "Alerting: Calculate diff for two AlertRules (#45877)" (#46023)"

This reverts commit 82aa5acba6b857d4eb7c6b5faf485ae6d20f7328.

* remove flakiness

alexweav committed:
Alerting: Add integration tests for dual-stage email templating (#43484)

* Resolve merge conflicts

* Remove cruft from local exploration

* Move integration tests to intercept using new abstraction layer instead of channel

* Fix linter error after rebase

grobinson-grafana committed:
Alerting: Fix use of > instead of >= when checking the For duration (#46011)

renovate[bot] committed:
Update dependency rollup to v2.68.0 (#45747)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update typescript-eslint monorepo to v5.13.0 (#45988)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Update dependency eslint-plugin-react to v7.29.2 (#45880)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

marefr committed:
Docs: Fix typo in Forward OAuth identity for the logged-in user section (#46043)

Fixes #45938

renovate[bot] committed:
Update dependency ts-node to v10.6.0 (#46045)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

ryantxu committed:
Loki: support live streaming (#42804)

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=44520) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=44520) before we can accept your contribution.<br/>**19** out of **20** committers have signed the CLA.<br/><br/>:white_check_mark: nathanrodman<br/>:white_check_mark: ryantxu<br/>:white_check_mark: ying-jeanne<br/>:white_check_mark: zuchka<br/>:white_check_mark: Jguer<br/>:white_check_mark: leeoniya<br/>:white_check_mark: CatPerry<br/>:white_check_mark: torkelo<br/>:white_check_mark: tolzhabayev<br/>:white_check_mark: srclosson<br/>:white_check_mark: kalleep<br/>:white_check_mark: yuri-tceretian<br/>:white_check_mark: timlevett<br/>:white_check_mark: alexweav<br/>:white_check_mark: achatterjee-grafana<br/>:white_check_mark: grafanabot<br/>:white_check_mark: JohnnyQQQQ<br/>:white_check_mark: marefr<br/>:white_check_mark: grobinson-grafana<br/>:x: renovate[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=44520) it.</sub>

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu committed:
use flat loki format

ryantxu committed:
use flat loki format

ryantxu commented:
@gabor -- I updated this so we produce a single flat frame for loki results.  This can be merged without hurting/affecting anything else, but would greatly simplify your effort to incorporate in other PRs

gabor commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,404 @@
+			}
+		}
+
+		valueField.Name = valueField.Labels["__name__"]
+		delete(valueField.Labels, "__name__")
```
this is prometheus specific, Loki does not have `__name__`. i would recommend just not doing it, and giving the field a fixed name, like "Value" or "value".  the caller then can customize tha name & labels.

ryantxu replied:
🤔 ideally we have the same path for loki+prometheus, I updated it so we set the name to "Value" when the label is missing.  I think this matches expected behavior if a "__name__" label somehow exists

gabor replied:
1. i see your point. personally i would still prefer to move these kind of customizations to the caller, because i feel like we should do as little "magic" as possible, and considering this is a very easy transform for the user..i'll think more about it :-) 
2. i went to check how this is done currently in Prometheus and Loki, and it is a little different from how this PR does it:
    - Prometheus does not remove the `__name__` label from the field-labels. Loki does remove  the `__name__` from the field-labels, but i THINK this is a bug: 
        - https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/result_transformer.ts#L294 
            - we mutate the object that we get from the caller, and to me this does not seem intentional
            - i think we should change this for grafana9
    - the rest is the same for both Prom & Loki:
        - `field.Name` is always `Value`
        - `field.Config.DisplayNameFromDS` is based on `__name__`
        - `frame.Name` is based on `__name__`

gabor commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,404 @@
+	rsp := &backend.DataResponse{}
+
+	for iter.ReadArray() {
+		timeField := data.NewFieldFromFieldType(data.FieldTypeTime, 0) // for now!
```
for streams-data the timefield has a name, could we name this too? either "Time" or "time", i don't know which is "more standard".

ryantxu replied:
👍 set it to "Time" (using the golang constant)

gabor commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,404 @@
+		}
+	}
+
+	frame := data.NewFrame("", labelsField, timeField, lineField, tsField)
```
could we adjust this so that the first is the `timeField` and second is the `lineField` (rest does not matter)?

it makes my life easier when both metric-and-logs frames start with the timefield and then the valuefield.

gabor replied:
hmm.. after seeing https://github.com/grafana/grafana/blob/main/public/app/features/live/data/StreamingDataFrame.ts#L212-L213 i assume we want labels-field to be first :) ignore this one then, i'll update my code to handle this case.

gabor reviewed (commented):
LGTM! there are a couple minor things, nothing major.

i did a test "migration" and i can easily consume this format in my code while matching the test-data-snapshots 👍 

i cannot review the Go code from the "is this written how go should be written" perspective, someone with Go knowledge should do that i think.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=44520) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=44520) before we can accept your contribution.<br/>**19** out of **20** committers have signed the CLA.<br/><br/>:white_check_mark: ying-jeanne<br/>:white_check_mark: leeoniya<br/>:white_check_mark: ryantxu<br/>:white_check_mark: CatPerry<br/>:white_check_mark: nathanrodman<br/>:white_check_mark: Jguer<br/>:white_check_mark: JohnnyQQQQ<br/>:white_check_mark: kalleep<br/>:white_check_mark: torkelo<br/>:white_check_mark: srclosson<br/>:white_check_mark: tolzhabayev<br/>:white_check_mark: zuchka<br/>:white_check_mark: grafanabot<br/>:white_check_mark: marefr<br/>:white_check_mark: alexweav<br/>:white_check_mark: achatterjee-grafana<br/>:white_check_mark: grobinson-grafana<br/>:white_check_mark: yuri-tceretian<br/>:white_check_mark: timlevett<br/>:x: renovate[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=44520) it.</sub>

Commit:
Merge remote-tracking branch 'origin/main' into prom-parser

Commit:
update name structure

ryantxu commented:
> i cannot review the Go code from the "is this written how go should be written" perspective, someone with Go knowledge should do that i think.

@toddtreece -- do you mind taking a look?

toddtreece commented:
@ryantxu yep, i'm definitely not a go expert, but i'll take a look tonight

toddtreece commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+
+				for _, f := range lookup {
+					if f.Len() != max {
+						f.Append("") // no matching label
```
i'm not following what this chunk of code does. is this used to even out the length of the flelds in the frame?

ryantxu replied:
the goal is to get all the fields the same length.  I updated the section to be a bit more clear :)

toddtreece replied:
ah ok 👍 i'm following now with the recent changes

toddtreece commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+	return rsp
+}
+
+// For consistent ordering read values to an aray not a map
```
```suggestion
// For consistent ordering read values to an array not a map
```

toddtreece commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+						if len(rsp.Frames) == 0 {
+							rsp.Frames = append(rsp.Frames, data.NewFrame(""))
+						}
+						rsp.Frames[0].Fields = append(rsp.Frames[0].Fields, f)
```
should this always be adding the field to the first frame?

ryantxu replied:
I updated the section so intent is more clear

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

ryantxu commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+		case "data":
+			rsp = readPrometheusData(iter)
+
+		// case "error":
```
I don't think so -- target is initially just loki, and that will hit a different path anyway.  For prometheus, we need more test results to know what to do here.

ryantxu commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+				}
+			}
+
+			fmt.Printf("[data] TODO, support stats: %v\n", v)
```
removed

ryantxu commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+		// case "warnings":
+		default:
+			v := iter.Read()
+			fmt.Printf("[ROOT] TODO, support key: %s / %v\n", l1Field, v)
```
replaced with an internal logf function -- this is really useful while debugging a single file, but just noise otherwise

ryantxu commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+					default:
+						iter.Skip()
+						frame.AppendNotices(data.Notice{
+							Severity: data.NoticeSeverityError,
```
I don't think so -- it calls `iter.Skip` so the JSON should be OK, and I think better to return a partial result than explode if we can

toddtreece replied:
makes sense, i've seen both approaches with parsing errors, so i wasn't sure if there was a common pattern

ryantxu commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+		}
+	}
+
+	return rsp
```
good catch -- this was the original implementation that made a frame for each label set -- but in loki that ends up being almost every line!

ryantxu commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+	// 1645228233
+	// 1645030246277587968
+	ss, _ := strconv.ParseInt(str[0:10], 10, 64)
+	ns, _ := strconv.ParseInt(str[10:], 10, 64)
```
I agree with your theory -- parsing once is better than twice!  I looks like we will have issues when the value is bigger than: math.MaxInt65=9223372036854775807

I updated the code so it does a faster pass for values before 2033

toddtreece replied:
are all loki string timestamps in nanoseconds? i think at some point i got confused and thought this needed to handle both second and nanosecond parsing

gabor replied:
when loki is returning logs-data (not metric-data), the timestamps are nanosecond-timestamps-as-string-in-json, like:
```
        "values": [
          [
            "1569266497240578000",
            "foo"
          ],
          [
            "1569266492548155000",
            "bar"
          ]
       ]
```

when loki is returning metric-data, the result is in prometheus-format, so the timestamps are second-timestamps-as-numbers-in-json, like:
```
    "values": [
          [
            1588889221,
            "137.95"
          ],
          [
            1588889221,
            "467.115"
          ],
          [
            1588889221,
            "658.8516666666667"
          ]
        ]
```

ryantxu committed:
hide fmt, improve time parsing

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=44520) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=44520) before we can accept your contribution.<br/>**19** out of **21** committers have signed the CLA.<br/><br/>:white_check_mark: ryantxu<br/>:white_check_mark: ying-jeanne<br/>:white_check_mark: nathanrodman<br/>:white_check_mark: Jguer<br/>:white_check_mark: leeoniya<br/>:white_check_mark: CatPerry<br/>:white_check_mark: tolzhabayev<br/>:white_check_mark: kalleep<br/>:white_check_mark: zuchka<br/>:white_check_mark: srclosson<br/>:white_check_mark: timlevett<br/>:white_check_mark: grafanabot<br/>:white_check_mark: torkelo<br/>:white_check_mark: JohnnyQQQQ<br/>:white_check_mark: yuri-tceretian<br/>:white_check_mark: marefr<br/>:white_check_mark: alexweav<br/>:white_check_mark: grobinson-grafana<br/>:white_check_mark: achatterjee-grafana<br/>:x: renovate[bot]<br/>:x: Ryan McKinley<br/><hr/>**Ryan McKinley** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=44520) it.</sub>

ryantxu committed:
Update pkg/util/converter/prom.go

Co-authored-by: Todd Treece <360020+toddtreece@users.noreply.github.com>

ryantxu commented on pkg/util/converter/prom.go:
```diff
@@ -0,0 +1,410 @@
+			switch resultType {
+			case "matrix":
+				rsp = readMatrixOrVector(iter)
+			case "vector":
```
yes -- If we had real sample responses, then it would be easy to implement.  I have only seen them described, but not a result that includes them

ryantxu committed:
more comments

ryantxu committed:
Merge branch 'prom-parser' of github.com:grafana/grafana into prom-parser

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=44520) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=44520) before we can accept your contribution.<br/>**19** out of **21** committers have signed the CLA.<br/><br/>:white_check_mark: ryantxu<br/>:white_check_mark: ying-jeanne<br/>:white_check_mark: nathanrodman<br/>:white_check_mark: CatPerry<br/>:white_check_mark: zuchka<br/>:white_check_mark: kalleep<br/>:white_check_mark: tolzhabayev<br/>:white_check_mark: Jguer<br/>:white_check_mark: leeoniya<br/>:white_check_mark: JohnnyQQQQ<br/>:white_check_mark: timlevett<br/>:white_check_mark: grafanabot<br/>:white_check_mark: srclosson<br/>:white_check_mark: torkelo<br/>:white_check_mark: grobinson-grafana<br/>:white_check_mark: alexweav<br/>:white_check_mark: marefr<br/>:white_check_mark: yuri-tceretian<br/>:white_check_mark: achatterjee-grafana<br/>:x: renovate[bot]<br/>:x: Ryan McKinley<br/><hr/>**Ryan McKinley** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=44520) it.</sub>

ryantxu committed:
fix lint

toddtreece reviewed (approved):
nice work, and thanks for adding all of the test coverage 💯

i think my only outstanding question at this point is around loki timestamp formats, but i don't think that's a blocker here since i am guessing that will be easier to confirm once it's integrated into the loki data source

ryantxu committed:
Merge remote-tracking branch 'origin/main' into prom-parser

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=44520) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=44520) before we can accept your contribution.<br/>**19** out of **21** committers have signed the CLA.<br/><br/>:white_check_mark: ying-jeanne<br/>:white_check_mark: ryantxu<br/>:white_check_mark: CatPerry<br/>:white_check_mark: nathanrodman<br/>:white_check_mark: zuchka<br/>:white_check_mark: kalleep<br/>:white_check_mark: leeoniya<br/>:white_check_mark: tolzhabayev<br/>:white_check_mark: Jguer<br/>:white_check_mark: torkelo<br/>:white_check_mark: timlevett<br/>:white_check_mark: JohnnyQQQQ<br/>:white_check_mark: grafanabot<br/>:white_check_mark: srclosson<br/>:white_check_mark: grobinson-grafana<br/>:white_check_mark: marefr<br/>:white_check_mark: yuri-tceretian<br/>:white_check_mark: alexweav<br/>:white_check_mark: achatterjee-grafana<br/>:x: renovate[bot]<br/>:x: Ryan McKinley<br/><hr/>**Ryan McKinley** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=44520) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=44520) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=44520) before we can accept your contribution.<br/>**19** out of **21** committers have signed the CLA.<br/><br/>:white_check_mark: ryantxu<br/>:white_check_mark: nathanrodman<br/>:white_check_mark: leeoniya<br/>:white_check_mark: CatPerry<br/>:white_check_mark: ying-jeanne<br/>:white_check_mark: srclosson<br/>:white_check_mark: tolzhabayev<br/>:white_check_mark: zuchka<br/>:white_check_mark: kalleep<br/>:white_check_mark: Jguer<br/>:white_check_mark: achatterjee-grafana<br/>:white_check_mark: alexweav<br/>:white_check_mark: timlevett<br/>:white_check_mark: JohnnyQQQQ<br/>:white_check_mark: torkelo<br/>:white_check_mark: yuri-tceretian<br/>:white_check_mark: grafanabot<br/>:white_check_mark: grobinson-grafana<br/>:white_check_mark: marefr<br/>:x: renovate[bot]<br/>:x: Ryan McKinley<br/><hr/>**Ryan McKinley** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=44520) it.</sub>
